### PR TITLE
Fix/enable repeated relate

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

As described on [DPE-982](https://warthogs.atlassian.net/browse/DPE-982).

*TL;DR*  app A is related, unrelated. It will fail to relate app A (or any other app).

## Solution

- Added several cleanups for peer data keys used for control the process on `relation_broken` events.
- Extract user addition configuration step out of event handler, so it can be called independent of original event sequence, allowing app to be re-related

*EDIT*:
- Also ditched embed test charm in favor of [test app](https://charmhub.io/mysql-test-app)

## Context

The implementation was that a given sequence of steps would happen as following:

1. [database_provides] App data to router
	- set `provides_data` key info from app @ on_database_requested event
2. call on_database_relation_joined with after `provides_data` [database_requires]
3. receive on backend database created [database_requires]
	- if `database-created` key
	- set app username/password
	- set `requires-application-data` key
	- set `database-created`
	- create application credentials in DB
4. bootstrap mysqlrouter @ peer_relation_changed [charm]
	- instance is waiting status && `database-created` && mysql-router bootstrap successful
	- set Active status
	- set `unit-bootstrapped` key in unit databag
	- add bootstrap count `num-units-bootstrapped`
5. set credentials for database relation @ peer_relation_changed [database_provides]
	- if not `credentials-shared` && `database-created` && `unit-bootstrapped`
	- set app-database-relation credentials/endpoints
	- set `credentials-shared` true

No cleanup of the `keys` was in place, which wouldn't allow subsequent relations infer it's status and re trigger necessary steps.  

[DPE-982]: https://warthogs.atlassian.net/browse/DPE-982?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by taurus-forever — APPROVED** [review]
Tnx!

File: src/relations/database_provides.py [hunk 1]
```diff
@@ -95,3 +107,36 @@ def _on_peer_relation_changed(self, _) -> None:
...
+        # application user cleanup when backend relation still in place
+        if backend_relation := self.charm.model.get_relation(DATABASE_REQUIRES_RELATION):
+            if app_data := self.charm.app_peer_data.get(MYSQL_ROUTER_REQUIRES_APPLICATION_DATA):
+                username = json.loads(app_data).get("username")
+
+                db_username = backend_relation.data[backend_relation.app].get("username")
+                db_password = backend_relation.data[backend_relation.app].get("password")
+                db_host, db_port = (
+                    backend_relation.data[backend_relation.app].get("endpoints").split(":")
+                )
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/relations/database_provides.py:128]: At this point (relation broken) if the keys didn't exist a failure would had occurred way earlier when first relating to either app or backend database.

[the remark this answers is not in the document]
Comment (paulomach) [reply src/relations/database_provides.py:128]: sure

**Comment by paulomach** [comment]
> question: is the `mysql-test-app` the high availability test app or the relations test app (in mysql-k8s-operator)?

@carlcsaposs-canonical , is the high availability test + relation app as in mysql-router(s) (i.e. use dataplatform libs interfaces)
We need asses what are other tests within our repos are doing and migrate them to the test app as we go.

>  should the name include which test app it is?

I would prefer to describe in the test app README.md

**Comment by paulomach** [comment]
> @paulomach to confirm, does that mean eventually we will only have one test app repo for all the test applications used in mysql charm tests?

Sounds like a plan to me (as long as it does not turn into a massive behemoth of test code that we have to maintain).
